sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CaseInsensitiveMap.scala

```diff
@@ -18,21 +18,25 @@
 package org.apache.spark.sql.catalyst.util
 
 /**
- * Builds a map in which keys are case insensitive
+ * Builds a map in which keys are case insensitive. Input map can be accessed for cases where
+ * case-sensitive information is also required.
  */
-class CaseInsensitiveMap(map: Map[String, String]) extends Map[String, String]
+class CaseInsensitiveMap(val caseSensitiveMap: Map[String, String]) extends Map[String, String]
   with Serializable {
 
-  val baseMap = map.map(kv => kv.copy(_1 = kv._1.toLowerCase))
+  val baseMap = caseSensitiveMap.map(kv => kv.copy(_1 = kv._1.toLowerCase))
 
   override def get(k: String): Option[String] = baseMap.get(k.toLowerCase)
 
   override def contains(k: String): Boolean = baseMap.contains(k.toLowerCase)
 
-  override def + [B1 >: String](kv: (String, B1)): Map[String, B1] =
-    baseMap + kv.copy(_1 = kv._1.toLowerCase)
+  override def +[B1 >: String](kv: (String, B1)): Map[String, B1] = {
+    new CaseInsensitiveMap(caseSensitiveMap + kv.copy(_2 = kv._2.asInstanceOf[String]))
```

Comment:
copy is unnecessary, but I do need the cast, otherwise getting compiler error :
Error:(34, 47) type mismatch;
 found   : (String, B1)
 required: (String, String)
    new CaseInsensitiveMap(caseSensitiveMap + kv )

I  am thinking of changing it to the following :
new CaseInsensitiveMap(caseSensitiveMap + kv.asInstanceOf[(String, String)])

Thank you  for the feedback. 
                                              ^